### ISSUE TITLE
Deep Dive deadline storage method

### ISSUE THREAD

**Issue by ferruzzi** [apache/airflow#43989]
Determine storage method (in the DAG object or in a DB table or ??? )
- Must be able to sort/search by "need-by" and remove by "dagrun_id"
- Must be able to allow alternative remove options for task_id or other options for future work

**Comment by ferruzzi** [comment]
Current thought:

A new table in the DB which has a 1:1 mapping for identifier:deadline, where identifier can be a dagrun_id, task_id, etc.  and a query which returns the deadlines filtered to only return the ones which has passed

**Comment by ferruzzi** [comment]
If we are only storing dagrun_id and task_id in the identifier column then it's easy to infer what type of id it is from the format, but does every possible future identifier type have a name format where we can infer what it is from the pattern?  Presumably not?  So we likely need a `type` column as well just to be safe?

**Comment by 1fanwang** [comment]
Sharing some of my early thoughts on the points raised here and from our offline discussion 

> "Where should deadlines be stored?"

Overall +1 to have it in a new/its own table for clear separation of concerns, this way we can also optimize for querying specifically for deadline, also this gives us the flexibility to support `identifier` that's mentioned here.

> Must be able to sort/search by "need-by"

We can have a dedicated index for `need-by`

> "Does every possible future identifier type have a name format where we can infer what it is?"

Maybe we can make this safe and explicit by having an `identifier_type` column instead of inferring from format?

> Must be able to remove by "dagrun_id"

Something like this? `DELETE FROM deadline_entry WHERE identifier_type = 'dagrun' AND identifier = 'dagrun_123'`

> Must be able to allow alternative remove options for future work

Assuming this is saying we might query based on different identifier types
`DELETE FROM deadline_entry WHERE dag_id = 'dag1' AND task_id = 'task1'`

> "1:1 mapping for identifier:deadline"

We can enforce unique constraint on (identifier_type, identifier)


WDYT

**Comment by ferruzzi** [comment]
Yeah, the more I think about this, the more I think I am on the right track.  I don't have much experience with db design so I want to bounce it off a couple people to verify before I/we commit to it, but yes, based on your replies I think this is the right approach.

A new table called deadlines with three columns:   id, id_type, and deadline which has a primary key on id so we can lookup and drop by id, and an index on deadline so we can easily get MIN(deadline).  `dag_processor` will calculate the deadline and add a row when the dagrun is created.  A call in the scheduler loop will query `SELECT MIN(deadline) AS earliest_deadline FROM deadlines;` and compare that to now() to see if any deadline has passed.   If there are any values returned, then call the DeadlinesHandler.

The DeadlinesHandler will get all deadlines that have passed, queue their callbacks, and drop them from the table so they are not queued again next pass.  We will need to make sure we can fetch the callback given the id and id_type OR store the callback in the table directly and save that lookup.  I think the callback will already be stored in the dagrun table so we should fetch from there.

Add logic to the dagrun cleanup/exit code to the effect of "if this dagrun has a Deadline then try to drop it from the table".  We "try" because if the deadline passed then it would have been dropped already.

---

I don't understand this part of the question:

```
Also curious on what we imagined the relationship would look like, quick gut check on my understanding of the relationship of dag dagrun task taskinstance
would it roughly look something like this?

DeadlineEntry -> DagRun: Optional reference
DeadlineEntry -> TaskInstance: Optional reference
DeadlineEntry -> DAG: Optional reference (via dag_id)
```

**Comment by 1fanwang** [comment]
> I don't understand this part of the question:

 ah sorry I was just thinking out loud on how the new Deadline table would relate to existing Airflow entities (DAG, DagRun, Task, and TaskInstance)
 
 I think I got the gist of it now after reading through your comment now, thanks!

**Comment by dstandish** [comment]
my advice @ferruzzi is don't spend too much time thinking about the model ahead of time.  I would make your primary focus the functionalities that you need to implement, and just make prs for those one at a time, starting from the simplest.  in process of code review, folks can help shape it along, and spot e.g. potential performance concerns etc.  when you try to just come up with the end state model in the beginning, inevitably it will have to be changed.  as the mike tyson famously said, everybody has a plan until they get punched in the face :)

**Comment by potiuk** [comment]
Very much agree. It's called "emerging design".
